src/python-common/ceph/rgw/rgwam_core.py

```diff
@@ -824,6 +828,11 @@ def zone_create(self, rgw_spec, start_radosgw, secondary_zone_period_retry_limit
                                 False,  # secondary zone
                                 access_key, secret, endpoints=rgw_spec.zone_endpoints)
 
+        # Make the newly created realm as default realm if it is the only realm present
```

Comment:
why is this part necessary?

Reply:
This is mentioned in the docs - https://docs.ceph.com/en/quincy/radosgw/multisite/#pulling-the-realm-configuration. Moreover as per the current implementation in dashboard, When we do realm pull on a cluster with a rgw service running on the initial `default` named zonegroup and zone, we store the rgw credentials for that realm in the dashboard config like - `{"realm1": "PFIR20LCC8KP126OB7ON"}`. Also we restart the rgw daemons in this process, so everytime we visit the buckets/users page it fetches the daemon credentials from this config using the realm name. If we do not make the first realm default, the existing rgw daemons running on `default` named zonegorup and zone are not able to fetch the access/secret keys from the config and throw 404 Credentials not found error because they are not included in any realm. If we make this realm default,on daemon restart the existing daemons running on `default` zg/zone get deployed under this realm and we can get credentials from the config.
If we already have realm/s in place and all daemons are running on any of the realm, then we do not see this issue.

Reply:
sorry, i'm not following the requirement to store things in the default zone/zonegroup or restart existing services

Reply:
In short, currently we do not allow a service to be running on the default zone/zonegroup if a realm is created in the setup (because we no longer store the access/secret keys of the user used to create that service), So we restart that service to make it run on newly created realm/zg/one by making the pulled realm default if it's the first one.

Reply:
> because we no longer store the access/secret keys of the user used to create that service

can you clarify this part? are we talking about the access/secret keys stored in the zone for rgw multisite? or the creds that the dashboard uses to access the rgw admin apis?

Reply:
The access/secret keys stored in the zone. We store these keys in a dashboard config like  - {"realm1": "PFIR20LCC8KP126OB7ON"}. So a rgw service running on default zg/zone doesn't have these keys stored in the config so it's unable to authenticate. This is all existing functionality. If we do not want to make the first realm as default then we need to change the way we store the zone keys in config and store the "default" zone's keys as well so the existing services running on this "default" zone are able to fetch these keys from the config.

Reply:
Till there is no realm, the "default" zone's keys are stored as plain strings (and not as a dict like I mentioned above which uses realm name to get the corresponding keys). When a realm is created we expect every service to be deployed on a realm so we can use service's realm name to fetch the keys from the dict

Reply:
> The access/secret keys stored in the zone. We store these keys in a dashboard config like - {"realm1": "PFIR20LCC8KP126OB7ON"}. So a rgw service running on default zg/zone doesn't have these keys stored in the config so it's unable to authenticate.

this makes it sound like you're referring to the "dashboard" user's credentials created by `ceph dashboard set-rgw-credentials`, which the dashboard uses to authenticate against rgw's admin apis. isn't that already configured and working for the default zone/zonegroup before we create any realms?

Reply:
Yes, the one's we get from `ceph dashboard set-rgw-credentials`. It works till there is no realm because we store the access/secret keys as plain strings because there are just one set of values to be used. When there is a realm/s we store the keys in a dict with realm name as keys and we donot store the "default" zone's keys in the config anymore because it has no realm and therefore it fails to authenticate

Reply:
if the admin has a default zone/zonegroup with no realm, they'll use commands like `radosgw-admin user create`, `radosgw-admin bucket rm` etc to operate on it. if we create a realm and set it as default, those commands will operate inside that realm instead, and won't do what the admin expects

the dashboard is meant to manage several independent rgw services. to maintain that independence, it shouldn't change or rely on any defaults

rgw intentionally changed that behavior in https://github.com/ceph/ceph/pull/59422/commits/d64e273de1623fe784a3b8c9d2c21990f78b7be2, so i'd prefer not to add it back here

Reply:
Thanks @cbodley , updated the PR again